Introduce a new VRF CNI meta plugin (also bump netlink package to v1.1.0)

fedepaol opened the pull request:
This plugin allows to create a [VRF](https://www.kernel.org/doc/Documentation/networking/vrf.txt) with the given name (or use the existing one if any) in the target namespace, and to allocate the interface to it.

VRFs make it possible to use multiple routing tables on the same namespace and allows isolation among interfaces within the same namespace. On top of that, this allows different interfaces to have overlapping CIDRs (or even addresses).

This is only useful in addition to other plugins.

The configuration is pretty simple and looks like:
```
{
    "type": "vrf",
    "vrfname": "blue"
}
```

## Note: 

In order to limit the size of the PR, additional tests will follow in a separate PR depending on this one.
Plus, another PR will be submitted to add an optional routingTableID parameter that will allow the user to specify the desired routing table ID.

Finally, a separate commit is added to upgrade the netlink version, which is required to be able to set VRFs as master (with the current version, that is possible only for bridges interfaces).

s1061123 commented:
You may need to add vrf in https://github.com/containernetworking/plugins/blob/master/plugins/linux_only.txt if vrf is only for linux.

s1061123 commented on plugins/meta/vrf/main.go:
```diff
@@ -0,0 +1,164 @@
+package main
```
Could you please add copyright section here?

fedepaol replied:
Done

s1061123 commented on plugins/meta/vrf/vrf.go:
```diff
@@ -0,0 +1,105 @@
+			}
+		}
+	}
+	return (maxTable + 1), nil
```
routing table id seems to be unsigned char, uint8 in golang. Could you please double check it?
https://github.com/torvalds/linux/blob/master/include/uapi/linux/rtnetlink.h#L227

fedepaol replied:
Sure, thanks for pointing out!

fedepaol replied:
@s1061123  I _think_ that's a u32.

See https://elixir.bootlin.com/linux/latest/source/drivers/net/vrf.c#L44
https://elixir.bootlin.com/linux/latest/source/drivers/net/vrf.c#L1339

Which also sticks with the type used by the netlink library: https://github.com/containernetworking/plugins/blob/53377543e1e91e984033e0af5e19bc9bb6973c65/vendor/github.com/vishvananda/netlink/link_linux.go#L2745

s1061123 commented on plugins/meta/vrf/vrf.go:
```diff
@@ -0,0 +1,105 @@
+	for _, l := range links {
+		if vrf, ok := l.(*netlink.Vrf); ok {
+			if vrf.Table > maxTable {
+				maxTable = vrf.Table
```
What is happen if all routing tables are consumed and there is no free routing table?

fedepaol replied:
Good point. Let's return an error in that case.

Now that I look back to it, probably the logic needs also to be strengthened a bit here, trying to look for the first free instead of taking the max + 1, especially considering that with a following PR users will be able to choose the tableid.

fedepaol replied:
Done

fedepaol commented:
> You may need to add vrf in https://github.com/containernetworking/plugins/blob/master/plugins/linux_only.txt if vrf is only for linux.

Thanks, I would have been looking for it today :-)

phoracek commented on plugins/meta/vrf/README.md:
```diff
@@ -0,0 +1,27 @@
+
+## Overview
+
+This plugin creates a [VRF](https://www.kernel.org/doc/Documentation/networking/vrf.txt) in the network namespace,and assigns it the interface passed in the arguments. If the VRF is already present in the namespace, it only adds the interface to it.
```
namespace and

fedepaol replied:
Done, thanks!

jellonek commented on plugins/meta/vrf/main.go:
```diff
@@ -0,0 +1,178 @@
+		if err != nil {
+			return err
+		}
+
```
So there is no removing of interfaces from VRF because we are expecting that (happy path) a next in chain plugin will destroy the interface, what will in fact also remove it from VRF?

IMO this should also cover unhappy path, so the interface will be left to next in chain plugin for del in the shape in which next plugin left it, with master interface set to what was before.

fedepaol replied:
Ah! You are right.
The only thing I am not sure of, is the part where I should restore the master, as the CNI has no memory.

@jellonek what if I unset the master, and let the ADD command return an error if the given interface already has one master set?

jellonek replied:
Yeah, this would be what i would expect. Good idea!

jellonek commented on plugins/meta/vrf/main.go:
```diff
@@ -0,0 +1,178 @@
+		// Meaning, we are deleting the last interface assigned to the VRF
+		if len(interfaces) == 1 && interfaces[0].Attrs().Name == args.IfName {
+			err = netlink.LinkDel(v)
+			if err != nil {
```
Please remove this `if` with its body or better ...

jellonek commented on plugins/meta/vrf/main.go:
```diff
@@ -0,0 +1,178 @@
+				return err
+			}
+		}
+		return err
```
... `return nil` explicitly there.

jellonek commented on plugins/meta/vrf/main.go:
```diff
@@ -0,0 +1,178 @@
+	}
+
+	err = ns.WithNetNSPath(args.Netns, func(_ ns.NetNS) error {
+		v, err := findVRF(conf.VRFName)
```
Please s/v/vrf/

jellonek commented on plugins/meta/vrf/main.go:
```diff
@@ -0,0 +1,178 @@
+		return err
+	}
+	err = ns.WithNetNSPath(args.Netns, func(_ ns.NetNS) error {
+		v, err := findVRF(conf.VRFName)
```
Please s/v/vrf/

jellonek commented on plugins/meta/vrf/main.go:
```diff
@@ -0,0 +1,178 @@
+	}
+
+	err = ns.WithNetNSPath(args.Netns, func(_ ns.NetNS) error {
+		v, err := findVRF(conf.VRFName)
```
Please s/v/vrf/

jellonek commented on plugins/meta/vrf/main.go:
```diff
@@ -0,0 +1,178 @@
+		if err != nil {
+			return err
+		}
+		ii, err := assignedInterfaces(v)
```
Please s/ii/vrfInterfaces/ or some other more verbose name than `ii`.

jellonek reviewed (commented):
Mostly small nits, but overall IMO looks good.

fedepaol commented:
@jellonek I _think_ I addressed all the comments. I went with the "no override master" policy, meaning that if the interface already has a master, the add will fail instead of overriding the current master. Please take another look.

fedepaol commented on plugins/meta/vrf/main.go:
```diff
@@ -0,0 +1,178 @@
+	bv "github.com/containernetworking/plugins/pkg/utils/buildversion"
+)
+
+// VrfNetConf represents the firewall configuration.
```
!!!!! That reveals where I "took inspiration from". Sorry I did not notice, changing it.

fedepaol replied:
Done

jellonek commented on plugins/meta/vrf/main.go:
```diff
@@ -0,0 +1,183 @@
+		return nil, nil, fmt.Errorf("configuration is expected to have a valid vrf name")
+	}
+
+	// Parse previous result.
```
Hmmm... Same comment is below, so this looks a bit suspicious as real parsing is done later, while there is only checking if it was provided in raw form.

fedepaol replied:
Ouch

fedepaol replied:
At least I was consistent in leaving the comments from where I copied the structure :laughing:  https://github.com/containernetworking/plugins/blob/0a1421a08cc909aa67360f749e90bab60206a01f/plugins/meta/firewall/firewall.go#L75

fedepaol replied:
Done

aojea commented on plugins/meta/vrf/README.md:
```diff
@@ -0,0 +1,27 @@
+It does not create any network interfaces and therefore does not bring connectivity by itself.
+It is only useful when used in addition to other plugins.
+
+## Configuration
```
can you make more detailed README with examples on how to use it with another plugins?
i.e. https://github.com/containernetworking/plugins/tree/master/plugins/meta/firewall

fedepaol replied:
Sure, will do!

aojea commented on plugins/meta/vrf/main.go:
```diff
@@ -0,0 +1,182 @@
+	err = ns.WithNetNSPath(args.Netns, func(_ ns.NetNS) error {
+		vrf, err := findVRF(conf.VRFName)
+
+		if _, ok := err.(netlink.LinkNotFoundError); ok {
+			vrf, err = createVRF(conf.VRFName)
+		}
+
+		if err != nil {
+			return err
+		}
```
should we check first for the erro and then if it is a LinkNotFound error?
```
		if err != nil {
		     if _, ok := err.(netlink.LinkNotFoundError); ok {
			vrf, err = createVRF(conf.VRFName)
                      }	 else {
                         return err
                     }
             }

fedepaol replied:
I don't like extra nesting, and the above version is behaviourally identical, as the type check will fail if err is nil.

But if you think this other version is more clear I can change.

aojea replied:
I was more curious about what happens if err=nil with this?
```err.(netlink.LinkNotFoundError);```
from your comments I assume is not a problem

aojea replied:
you are right, there is lot of nesting and is very ugly to read :)

aojea commented on plugins/meta/vrf/main.go:
```diff
@@ -0,0 +1,182 @@
+	err = ns.WithNetNSPath(args.Netns, func(_ ns.NetNS) error {
+		vrf, err := findVRF(conf.VRFName)
+		if err != nil {
+			return err
```
@dcbw is not the DEL idempotent? should it return nil ?

fedepaol replied:
Ah, maybe only if err is a `LinkNotFoundError`?

aojea replied:
you need a more CNI guy here, but the spec says

> Plugins should generally complete a DEL action without error even if some resources are missing. For example, an IPAM plugin should generally release an IP allocation and return success even if the container network namespace no longer exists, unless that network namespace is critical for IPAM management. While DHCP may usually send a 'release' message on the container network interface, since DHCP leases have a lifetime this release action would not be considered critical and no error should be returned. For another example, the bridge plugin should delegate the DEL action to the IPAM plugin and clean up its own resources (if present) even if the container network namespace and/or container network interface no longer exist.

https://github.com/containernetworking/cni/blob/master/SPEC.md

fedepaol replied:
Thanks for the explanation!
Yes, I will modify the behavior so the DEL fails only if there's a real failure and not because the resources are not there.

bboreham commented:
Do you know anything about the CI failures?
```
/home/travis/gopath/src/github.com/containernetworking/plugins/plugins/meta/vrf/vrf_test.go:32
  passes prevResult through unchanged [It]
  /home/travis/gopath/src/github.com/containernetworking/plugins/plugins/meta/vrf/vrf_test.go:80
  Unexpected error:
      <*errors.errorString | 0xc000085c20>: {
          s: "cmdAdd failed: could not add VRF vrf0: operation not supported",
      }
      cmdAdd failed: could not add VRF vrf0: operation not supported
  occurred
  /home/travis/gopath/src/github.com/containernetworking/plugins/plugins/meta/vrf/vrf_test.go:96
```

fedepaol commented:
> Do you know anything about the CI failures?
> 
> ```
> /home/travis/gopath/src/github.com/containernetworking/plugins/plugins/meta/vrf/vrf_test.go:32
>   passes prevResult through unchanged [It]
>   /home/travis/gopath/src/github.com/containernetworking/plugins/plugins/meta/vrf/vrf_test.go:80
>   Unexpected error:
>       <*errors.errorString | 0xc000085c20>: {
>           s: "cmdAdd failed: could not add VRF vrf0: operation not supported",
>       }
>       cmdAdd failed: could not add VRF vrf0: operation not supported
>   occurred
>   /home/travis/gopath/src/github.com/containernetworking/plugins/plugins/meta/vrf/vrf_test.go:96
> ```

Yep, vrf support is introduced in ubuntu bionic. For this reason I opened this other pr https://github.com/containernetworking/plugins/pull/526 (probably I should have cross-referenced them).
That introduced a couple of other issues as mentioned in the pr itself.

bboreham reviewed (approved):
LGTM, thanks for all the work

fedepaol commented:
> LGTM, thanks for all the work

It was a pleasure :-)
